shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -154,21 +189,56 @@ public void setMessageHandler(
       @Nullable TaskQueue taskQueue) {
     if (handler == null) {
       Log.v(TAG, "Removing handler for channel '" + channel + "'");
-      messageHandlers.remove(channel);
-    } else {
-      DartMessengerTaskQueue dartMessengerTaskQueue = null;
-      if (taskQueue != null) {
-        dartMessengerTaskQueue = createdTaskQueues.get(taskQueue);
-        if (dartMessengerTaskQueue == null) {
-          throw new IllegalArgumentException(
-              "Unrecognized TaskQueue, use BinaryMessenger to create your TaskQueue (ex makeBackgroundTaskQueue).");
-        }
+      synchronized (handlersLock) {
+        messageHandlers.remove(channel);
       }
-      Log.v(TAG, "Setting handler for channel '" + channel + "'");
+      return;
+    }
+    DartMessengerTaskQueue dartMessengerTaskQueue = null;
+    if (taskQueue != null) {
+      dartMessengerTaskQueue = createdTaskQueues.get(taskQueue);
+      if (dartMessengerTaskQueue == null) {
+        throw new IllegalArgumentException(
+            "Unrecognized TaskQueue, use BinaryMessenger to create your TaskQueue (ex makeBackgroundTaskQueue).");
+      }
+    }
+    Log.v(TAG, "Setting handler for channel '" + channel + "'");
+
+    LinkedList<BufferedMessageInfo> list;
+    synchronized (handlersLock) {
       messageHandlers.put(channel, new HandlerInfo(handler, dartMessengerTaskQueue));
+      if (!bufferedMessages.containsKey(channel)) {
+        return;
+      }
+      list = (LinkedList) bufferedMessages.get(channel);
```

Comment:
Remove the `get` and just call `remove`.  `Map.remove` will return the removed value (if any)

Also change `list` to a `List<BufferedMessageInfo>` to match the value type of `bufferedMessages`.  And this cast should use the `List<BufferedMessageInfo>` parameterized type.

Reply:
nice